### PR TITLE
Datemath: Support fiscal years

### PR DESCRIPTION
**Pull request by sakjur**
**What this PR does / why we need it**:

Time range parsing in the Grafana backend currently doesn't support fiscal years and quarters which the frontend supports.

I also refactored the time range parsing to use a functional options style instead of adding yet another `ParseXWithY` family of methods.

**Which issue(s) this PR fixes**:

Partially fixing support for quarters and financial quarters for reporting.

- https://github.com/grafana/grafana-enterprise/issues/2062

**Special notes for your reviewer**:

I noted that this package is marked as deprecated in favor of [grafana/grafana-plugin-sdk-go](https://github.com/grafana/grafana-plugin-sdk-go/tree/main/backend), but that doesn't seem to contain the time range logic, so it's not a useful replacement for this particular structure.

I've specifically not refactored this part of this package away from the deprecated package (where it doesn't seem to belong) because I'm unsure where to put it.

### CODE REVIEW

File: pkg/tsdb/legacydata/time_range.go [hunk 1]
```diff
@@ -21,99 +21,167 @@ func NewDataTimeRange(from, to string) DataTimeRange {
...
-	return parseTimeRange(tr.From, tr.Now, false, location)
+func (tr DataTimeRange) ParseFromWithWeekStart(location *time.Location, weekstart time.Weekday) (time.Time, error) {
+	return tr.ParseFrom(WithLocation(location), WithWeekstart(weekstart))
 }
 
-func (tr DataTimeRange) ParseToWithLocation(location *time.Location) (time.Time, error) {
-	return parseTimeRange(tr.To, tr.Now, true, location)
+func (tr DataTimeRange) ParseToWithWeekStart(location *time.Location, weekstart time.Weekday) (time.Time, error) {
+	return tr.ParseTo(WithLocation(location), WithWeekstart(weekstart))
 }
```
Comment (sakjur) [inline pkg/tsdb/legacydata/time_range.go:84]: Without keeping these methods, we'd break Grafana Enterprise on merge. They can be removed once Grafana Enterprise has been updated to use the new functional options.

**Review by AgnesToulet — APPROVED** [review]
LGTM! I'd be interested to have @grafana/backend-platform opinion on when / if we should plan to move this part of the backend in [grafana/grafana-plugin-sdk-go](https://github.com/grafana/grafana-plugin-sdk-go/tree/main/backend) but I think this is out of scope for this PR.

**Comment by sakjur** [comment]
There's a known off-by-one bug attached to this that I've submitted a fix in upstream for:

- https://github.com/vectordotdev/go-datemath/pull/8

**Comment by sakjur** [comment]
Pushed this to 8.4 instead with the hope to include the upstream fix. Might move it to an earlier v8.3.x build if that goes quickly.

**Pull request merged**